File: LeetCode Solutions/Problems/732. My Calendar III.py
from sortedcontainers import SortedDict
class MyCalendarThree:

    def __init__(self):
        self.calendar = SortedDict()

    def book(self, start: int, end: int) -> int:
        if start in self.calendar:
            self.calendar[start] += 1
        else:
            self.calendar[start] = 1
        if end in self.calendar:
            self.calendar[end] += -1
        else:
            self.calendar[end] = -1
        counter, maxNo = 0, 0
        for key in self.calendar.keys():
            counter += self.calendar[key]
            maxNo = max(maxNo,counter)
        return maxNo


# Your MyCalendarThree object will be instantiated and called as such:
# obj = MyCalendarThree()
# param_1 = obj.book(start,end)

File: LeetCode Solutions/Problems/732. My Calendar III.java
class MyCalendarThree {
    TreeMap<Integer, Integer> calendar;
    public MyCalendarThree() {
        calendar = new TreeMap<>();
    }
    
    public int book(int start, int end) {
        calendar.put(start, calendar.getOrDefault(start, 0)+1);
        calendar.put(end, calendar.getOrDefault(end, 0)-1);
        int maxNo = 0, counter = 0;
        for (int val: calendar.values()){
            maxNo = Math.max(maxNo, counter += val);
        }
        return maxNo;
    }
}

/**
 * Your MyCalendarThree object will be instantiated and called as such:
 * MyCalendarThree obj = new MyCalendarThree();
 * int param_1 = obj.book(start,end);
 */